File: platform/avalonia/DDPlayer/AvoxCommon/SourcePlayer.cs
using AvoxCommon;
using AvoxNet;
using System;

namespace AvoxCommon
{
    /// <summary>
    /// SourcePlayer 封装类
    /// 针对 ISourcePlayer 接口的封装，用于处理直出的设备源（摄像头、麦克风等）
    /// 不需要解码器，也不需要音视频同步
    /// 注意：SourcePlayer 使用 AsrMode::streaming 模式，不支持翻译
    /// </summary>
    public class SourcePlayer : BasePlayer, IDisposable
    {
        private AvoxNet.ISourcePlayer sourcePlayer;
        private ISubtitle? subtitle;
        private bool isDisposed = false;

        /// <summary>
        /// 底层 ISourcePlayer 实例
        /// </summary>
        public AvoxNet.ISourcePlayer Player => sourcePlayer;

        /// <summary>
        /// 构造函数
        /// </summary>
        public SourcePlayer()
        {
            sourcePlayer = AvoxWrapper.createDevicePlayer();
            AvoxWrapper.addSourcePlayerOb(sourcePlayer, this);
        }

        /// <summary>
        /// 获取视频渲染器
        /// </summary>
        public ISurfaceRender GetSurfaceRender()
        {
            return sourcePlayer?.getSurfaceRender();
        }

        /// <summary>
        /// 获取音频渲染器
        /// </summary>
        public IAudioRender GetAudioRender()
        {
            return sourcePlayer?.getAudioRender();
        }

        /// <summary>
        /// 配置语音识别
        /// SourcePlayer 使用 AsrMode::streaming 模式（不支持翻译）
        /// </summary>
        /// <param name="config">STT 配置</param>
        public void EnableSttWithRender(SttConfig config)
        {
            try
            {
                subtitle = sourcePlayer?.getSubtitle();
                if (subtitle == null)
                {
                    AvoxWrapper.logMsg(LogLevel.warn, "Subtitle 为空，无法启用语音识别");
                    return;
                }
                // 设置字体样式（通过 IFontLayer）
                var render = sourcePlayer?.getSurfaceRender();
                if (render != null)
                {
                    var fontLayer = AvoxWrapper.enableRenderFont(render);
                    if (fontLayer != null)
                    {
                        fontLayer.setScale(config.FontScale);
                        fontLayer.setColor(config.FontColorR, config.FontColorG, config.FontColorB, config.FontOpacity);
                    }
                }
                // 启用 ASR (SourcePlayer 内部固定使用 streaming 模式)
                subtitle.enableAsr();
                AvoxWrapper.logMsg(LogLevel.info, $"SourcePlayer 语音识别已启用: AsrMode=streaming");
            }
            catch (Exception ex)
            {
                AvoxWrapper.logMsg(LogLevel.error, $"启用语音识别失败: {ex.Message}");
            }
        }

        /// <summary>
        /// 禁用语音识别
        /// </summary>
        public void DisableSttWithRender()
        {
            try
            {
                if (subtitle != null)
                {
                    subtitle.close();
                    subtitle.Dispose();
                    subtitle = null;
                }
                AvoxWrapper.logMsg(LogLevel.info, "SourcePlayer 语音识别已禁用");
            }
            catch (Exception ex)
            {
                AvoxWrapper.logMsg(LogLevel.error, $"禁用语音识别失败: {ex.Message}");
            }
        }

        /// <summary>
        /// 设置音频源
        /// </summary>
        public void SetAudioSource(IAudioSource source)
        {
            try
            {
                sourcePlayer?.setAudioSource(source);
            }
            catch (Exception ex)
            {
                AvoxWrapper.logMsg(LogLevel.error, $"设置音频源失败: {ex.Message}");
            }
        }

        /// <summary>
        /// 设置视频源
        /// </summary>
        public void SetVideoSource(IVideoSource source)
        {
            try
            {
                sourcePlayer?.setVideoSource(source);
            }
            catch (Exception ex)
            {
                AvoxWrapper.logMsg(LogLevel.error, $"设置视频源失败: {ex.Message}");
            }
        }

        /// <summary>
        /// 获取复用器（用于录制或推流）
        /// </summary>
        public IMediaMuxer GetMuxer()
        {
            return sourcePlayer?.getMuxer();
        }

        /// <summary>
        /// 打开源播放
        /// </summary>
        public bool Open()
        {
            try
            {
                bool result = sourcePlayer?.open() ?? false;
                if (result)
                {
                    var sttConfig = ConfigManager.Instance.Config.Stt;
                    if (sttConfig?.EnableStt == true)
                    {
                        EnableSttWithRender(sttConfig);
                    }
                }
                return result;
            }
            catch (Exception ex)
            {
                AvoxWrapper.logMsg(LogLevel.error, $"源播放打开失败: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// 关闭源播放
        /// </summary>
        public void Close()
        {
            DisableSttWithRender();
            sourcePlayer?.close();
            AvoxWrapper.logMsg(LogLevel.info, "源播放关闭");
        }

        /// <summary>
        /// 获取当前状态
        /// </summary>
        public PlayerState GetState()
        {
            return sourcePlayer?.getState() ?? PlayerState.none;
        }

        /// <summary>
        /// 获取源信息
        /// </summary>
        public ISourceInfo GetSourceInfo()
        {
            return sourcePlayer?.getSourceInfo();
        }

        /// <summary>
        /// 重写状态变更，同步底层状态
        /// </summary>
        public override void onStateChange(PlayerState preState, PlayerState state)
        {
            base.onStateChange(preState, state);
        }

        /// <summary>
        /// 释放资源
        /// </summary>
        public override void Dispose()
        {
            if (!isDisposed)
            {
                isDisposed = true;
                if (sourcePlayer != null)
                {
                    AvoxWrapper.removeSourcePlayerOb(sourcePlayer, this);
                    sourcePlayer.close();
                    sourcePlayer.Dispose();
                    sourcePlayer = null;
                }
                base.Dispose();
                AvoxWrapper.logMsg(LogLevel.info, "SourcePlayer 资源释放完成");
            }
        }
    }
}

File: src/avox/player/SourcePlayer.cpp
#include "SourcePlayer.hpp"

#include "../module/AvoxManager.hpp"

namespace avox {

SourcePlayer::SourcePlayer() {
  windowRender = std::make_unique<WindowRender>();
  audioRender = std::unique_ptr<AudioOutput>(getDefaultAudioOutput());
  rawMuxer = std::make_unique<RawMuxer>();
  source = std::make_unique<DeviceSource>();
  subtitleView = std::make_unique<SubtitleView>();
  subtitleView->setAsrMode(AsrMode::streaming);
  audioRender->enableOutput(false);
  mpCommands.setMaxSize(100);
  // 专门用来处理改变播放状态的线程
  startTask();
}

SourcePlayer::~SourcePlayer() {
  LOGFLF(LogLevel::info, "destroy source player");
  stopTask();
}

ISurfaceRender* SourcePlayer::getSurfaceRender() { return windowRender.get(); }

IAudioRender* SourcePlayer::getAudioRender() { return audioRender.get(); }

ISubtitle* SourcePlayer::getSubtitle() { return subtitleView.get(); }

void SourcePlayer::setAudioSource(IAudioSource* device) {
  source->setAudioSource(device);
}

void SourcePlayer::setVideoSource(IVideoSource* device) {
  source->setVideoSource(device);
}

IMediaMuxer* SourcePlayer::getMuxer() { return rawMuxer.get(); }

bool SourcePlayer::open() {
  // rawMuxer在源可用后,才能录制
  rawMuxer->setContext(this);
  // 转到线程上执行
  auto openCmd = createCommand<MPCommandType::Open>("");
  mpCommands.enqueueWait(openCmd);
  return true;
}

void SourcePlayer::close() {
  // 需要把之前的命令清空吗？这样可以快速关闭
  mpCommands.clear();
  // 停止
  auto stopCmd = createCommand<MPCommandType::Close>();
  mpCommands.enqueueWait(stopCmd);
}

PlayerState SourcePlayer::getState() { return state; }

void SourcePlayer::onRunTask() {
  while (running()) {
    MPCommandPtr cmd = nullptr;
    if (mpCommands.dequeue(cmd)) {
      switch (cmd->type) {
        case MPCommandType::Open: {
          // open,得到IO回复变成ready
          cmdOpen();
          break;
        }
        case MPCommandType::Ready: {
          // MediaPlayer::onOpen后得到流信息
          // 在打开解码线程后，进入opening状态
          cmdReady();
          break;
        }
        case MPCommandType::Close: {
          // stop
          cmdClose();
          break;
        }
        default:
          break;
      }
    }
    sleepTask(false, 10);
  }
  if (source) {
    source->close();
    source->removeObserver(this);
  }
}

void SourcePlayer::cmdOpen() {
  if (!source) {
    LOGFLF(LogLevel::warn, "source is null");
    return;
  }
  // 如果播放器已经打开,先清空资源
  if (state != PlayerState::none && state != PlayerState::stopped) {
    // 先关闭之前对象
    cmdClose();
  }
  source->setObserver(this);
  windowRender->addObserver(this);
  windowRender->start();
  bool bOpen = source->open();
  if (bOpen) {
    setState(PlayerState::opening);
  }
}

void SourcePlayer::cmdReady() {
  if (!source) {
    return;
  }
  auto vtracks = source->getVideoTracks();
  auto atracks = source->getAudioTracks();
  if (vtracks.size() > 0) {
    VideoDesc vdesc = vtracks[0].desc;
    LOGFLF(LogLevel::info, "video track: ", vdesc);
    subtitleView->setWindowRender(windowRender.get());
  }
  if (atracks.size() > 0) {
    AudioDesc adesc = atracks[0].desc;
    LOGFLF(LogLevel::info, "audio track: ", adesc);
    audioRender->setDesc(adesc);
    subtitleView->setAudioDesc(adesc);
    bAudioRender = true;
  }
  setState(PlayerState::ready);
  MPOB::dispatch(&IMediaPlayerOb::onReady);
}

void SourcePlayer::cmdClose() {
  if (source) {
    source->close();
    source->removeObserver(this);
  }
  // 如果先关渲染再关资源时有可能导致android surface出错
  // 这样下次android相机打开可能就不能正常显示
  windowRender->stop();
  windowRender->removeObserver(this);
  subtitleView->close();
  setState(PlayerState::stopped);
  MPOB::dispatch(&IMediaPlayerOb::onClose);
}

ISourceInfo* SourcePlayer::getSourceInfo() { return source.get(); }

void SourcePlayer::onMuxerOpen(MediaMuxer* muxer) {
  auto vtracks = source->getVideoTracks();
  auto atracks = source->getAudioTracks();
  if (vtracks.size() > 0) {
    bool bHard = rawMuxer->getHardEncode();
    YuvType outYuv = windowRender->getOutYuv();
    // 如果自身输出YUV420P,关闭硬解
    if (outYuv == YuvType::yuv420P) {
      bHard = false;
      rawMuxer->setHardEncode(false);
    }
    VTrackDesc vdesc = vtracks[0];
    // 录制颜色空间: shader 矩阵与 encoder tag 的共同源(阶段3 量程分流在此切换)
    ColorSpaceDesc cs{YuvStandard::bt601, YuvRange::full};
    vdesc.desc.colorSpace = cs;
    // 同源驱动 shader 矩阵(rgba2YUV)
    windowRender->setColorSpace(cs);
    if (bHard) {
#ifdef WIN32
      // 如果没打开YUV输出
      if (outYuv == YuvType::other) {
        windowRender->enableYuvOut(YuvType::nv12);
        bMuxerCpu = true;
      }
#endif
      vdesc.desc.type = YuvType::nv12;
    } else {
      // ffmpeg软解需要Yuv420P
      windowRender->enableYuvOut(YuvType::yuv420P);
      vdesc.desc.type = YuvType::yuv420P;
    }
    muxer->setInVideoDesc(vdesc);
  }
  if (atracks.size() > 0) {
    muxer->setInAudioDesc(atracks[0]);
  }
  muxer->ready();
}

void SourcePlayer::onMuxerClose() {
  // 如果是因为录制打开的CPU模式，现在需要关闭
  if (bMuxerCpu) {
    windowRender->disableYuvOut();
    bMuxerCpu = false;
  }
}

void SourcePlayer::onReady() {
  auto readyCmd = createCommand<MPCommandType::Ready>();
  mpCommands.enqueueWait(readyCmd);
}

void SourcePlayer::onError(AVError error, const char* msg) {
  if (error == AVError::none) {
    return;
  }
  if (error == AVError::endOfFile) {
    setState(PlayerState::completed);
    MPOB::dispatch(&IMediaPlayerOb::onComplete);
  } else {
    if (error == AVError::deviceDisable) {
    }
    LOGFLF(LogLevel::warn, "raw source error:", getAVErrorStr(error),
           " msg:", msg);
    MPOB::dispatch(&IMediaPlayerOb::onIoError, error, msg);
  }
}

void SourcePlayer::onVideoFrame(const YUVFrame& frame, int32_t trackId) {
  if (state == PlayerState::ready) {
    setState(PlayerState::playing);
  }
  windowRender->render(frame);
  if (rawMuxer->getState() == RecorderState::recording) {
    windowRender->pushFrame(rawMuxer.get());
  }
}

void SourcePlayer::onGpuFrame(const GpuFrame& frame, int32_t trackId) {
  if (state == PlayerState::ready) {
    setState(PlayerState::playing);
  }
  // HighClock clock = {};
  windowRender->render(frame);
  if (rawMuxer->getState() == RecorderState::recording) {
    windowRender->pushFrame(rawMuxer.get());
  }
}

void SourcePlayer::onAudioFrame(const AvoxAFrame& frame, int32_t trackId) {
  if (state == PlayerState::ready) {
    setState(PlayerState::playing);
  }
  if (audioRender) {
    audioRender->render(frame.buffer, frame.pts);
  }
  if (subtitleView) {
    subtitleView->inputSpeech(frame.buffer, frame.pts);
  }
  if (rawMuxer->getState() == RecorderState::recording) {
    rawMuxer->pushFrame(frame);
  }
}

void SourcePlayer::onClose() { subtitleView->close(); }

}
